### PR TITLE
add crop layer

### PR DESCRIPTION
**Pull request by wanghaoshuang**
solve #2470

### CODE REVIEW

**Comment by qingqing01** [comment]
Cropping and Padding are the opposite operations. Can we use the padding function for this cropping operation?

**Comment by wanghaoshuang** [comment]
@qingqing01 @pkuyym  padding function and cropping funtion is not opposite operations, strictly?
Padding forward funtion pad zero(or other values) to output tensor, while cropping backward function won`t change marginal element of output tensor.

'Padding forward funtion' != 'Cropping backward function'

**Comment by wanghaoshuang** [comment]
@qingqing01 @pkuyym Thks for interpretion and discussion about cropping function and padding function. I think i got it.

File: proto/ModelConfig.proto [hunk 1]
```diff
@@ -266,6 +266,12 @@ message PadConfig {
   repeated uint32 pad_w = 4;
 }
 
+message CropConfig {
+  required ImageConfig image_conf = 1;
+  repeated uint32 crop_corner = 2;
+  repeated uint32 crop_shape = 3;
+}
```
Comment (qingqing01) [inline proto/ModelConfig.proto]: Both Caffe and MXNet use `offset` to indicate `crop_corner`.

File: paddle/gserver/layers/CropLayer.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  int h = inputLayers_[0]->getOutput().getFrameHeight();
+  if (h != 0) inDims_.setDim(2, h);
+  int w = inputLayers_[0]->getOutput().getFrameWidth();
+  if (w != 0) inDims_.setDim(3, w);
+  setOutDims(batchSize);
+}
+
+void CropLayer::forward(PassType passType) {
+  Layer::forward(passType);
+  MatrixPtr input = inputLayers_[0]->getOutputValue();
```
Comment (qingqing01) [inline paddle/gserver/layers/CropLayer.cpp]: This layer should be able to accept two inputs, get the height and width from the second input.

File: proto/ModelConfig.proto (hunk 1 above)
Comment (wanghaoshuang) [inline proto/ModelConfig.proto]: Thx，get it.

File: paddle/gserver/layers/CropLayer.cpp (hunk 2 above)
Comment (wanghaoshuang) [inline paddle/gserver/layers/CropLayer.cpp]: Get it.

**Comment by qingqing01** [comment]
@wanghaoshuang 如果提交了comments，请回复下哈~

File: paddle/gserver/layers/CropLayer.cpp (hunk 2 above)
Comment (wanghaoshuang) [inline paddle/gserver/layers/CropLayer.cpp]: @qingqing01 I have submit a pr to add python wrapper and grad test for crop layer. But TeamCity build job was blocked by issue #2745

File: python/paddle/trainer_config_helpers/layers.py [hunk 3]
```diff
@@ -5876,3 +5877,56 @@ def prelu_layer(input,
...
+    :param shape: The shape to be cropped. Default is None.
+    :type shape: Sqquence | None
+    :param name: Name of this layer.
+    :type name: basestring
+    :return: LayerOutput object.
+    :rtype: LayerOutput
+    """
+    if isinstance(input, LayerOutput):
+        input = [input]
+    elif isinstance(input, Projection):
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: 不支持`Projection` 输入，应该去掉。

File: python/paddle/trainer_config_helpers/layers.py [hunk 4]
```diff
@@ -5876,3 +5877,56 @@ def prelu_layer(input,
         layer_type=LayerType.PRELU,
         parents=input,
         size=l.config.size)
+
+
+@wrap_name_default()
+@layer_support()
+def crop_layer(input, axis, offset, shape=None, name=None, layer_attr=None):
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: axis给个默认值吧。

File: python/paddle/trainer_config_helpers/layers.py [hunk 5]
```diff
@@ -5876,3 +5877,56 @@ def prelu_layer(input,
...
+    :param axis: start axis to be cropped. To image input layer:
+        - 0: batch size
+        - 1: channels
+        - 2: height
+        - 3: width
+    :type partial_sum: int
+    :param offset: The crop offset
+    :type offset: Sequence
+    :param shape: The shape to be cropped. Default is None.
+    :type shape: Sqquence | None
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: Sqquence是啥？

File: python/paddle/trainer_config_helpers/layers.py [hunk 6]
```diff
@@ -5876,3 +5877,56 @@ def prelu_layer(input,
         layer_type=LayerType.PRELU,
         parents=input,
         size=l.config.size)
+
+
+@wrap_name_default()
+@layer_support()
+def crop_layer(input, axis, offset, shape=None, name=None, layer_attr=None):
+    """
+    The crop layer crop images by offset and shape. User can set crop shape by
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: The crop layer crop -> The crop layer crops

File: paddle/gserver/layers/CropLayer.cpp [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+void CropLayer::setOutDims(const size_t batchSize) {
+  outDims_.reshape({batchSize, crop_shape_[1], crop_shape_[2], crop_shape_[3]});
+}
+
+void CropLayer::setTensorDim(const size_t batchSize) {
+  CHECK_EQ(static_cast<int>(inputLayers_.size()), 2);
+  inDims_.setDim(0, batchSize);
+  int h = inputLayers_[0]->getOutput().getFrameHeight();
+  if (h != 0) inDims_.setDim(2, h);
+  int w = inputLayers_[0]->getOutput().getFrameWidth();
```
Comment (qingqing01) [inline paddle/gserver/layers/CropLayer.cpp]: 应该是：inputLayers_[1]？看Python接口里，第2个input是reference_input

File: paddle/gserver/layers/CropLayer.cpp [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+  return true;
+}
+
+void CropLayer::setOutDims(const size_t batchSize) {
+  outDims_.reshape({batchSize, crop_shape_[1], crop_shape_[2], crop_shape_[3]});
+}
+
+void CropLayer::setTensorDim(const size_t batchSize) {
+  CHECK_EQ(static_cast<int>(inputLayers_.size()), 2);
```
Comment (qingqing01) [inline paddle/gserver/layers/CropLayer.cpp]: Python接口写的是支持一个或两个输入，这块却是`CHECK_EQ` ?

File: python/paddle/trainer_config_helpers/layers.py (hunk 4 above)
Comment (wanghaoshuang) [inline python/paddle/trainer_config_helpers/layers.py]: fixed

File: python/paddle/trainer_config_helpers/layers.py (hunk 6 above)
Comment (wanghaoshuang) [inline python/paddle/trainer_config_helpers/layers.py]: fixed

File: python/paddle/trainer_config_helpers/layers.py (hunk 5 above)
Comment (wanghaoshuang) [inline python/paddle/trainer_config_helpers/layers.py]: Fixed. Sqquence->Sequence

File: python/paddle/trainer_config_helpers/layers.py (hunk 3 above)
Comment (wanghaoshuang) [inline python/paddle/trainer_config_helpers/layers.py]: Fixed.

File: paddle/gserver/layers/CropLayer.cpp (hunk 8 above)
Comment (wanghaoshuang) [inline paddle/gserver/layers/CropLayer.cpp]: Fixed.

File: paddle/gserver/layers/CropLayer.cpp (hunk 7 above)
Comment (wanghaoshuang) [inline paddle/gserver/layers/CropLayer.cpp]: 这里是取原始input的dims，所以是inputLayers_[0].

**Comment by wanghaoshuang** [comment]
@qingqing01 crop layer相关bugs已经fixed.已commit.

**Pull request merged**